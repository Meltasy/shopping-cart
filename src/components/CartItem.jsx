const CartItem = ({ itemId, price, quantity, productItems, handleDelete }) => {
  const item = productItems.find(item => item.id === itemId)
  const totalCost = price * quantity
  if (item == null) return null

  return (
    <div key={item.id} className='itemCart'>
      <img src={item.image} alt=' ' width={100} height={100}/>
      <div>
        <h4>{item.title} €{price}</h4>
        <h3>x{quantity} €{totalCost}</h3>
      </div>
      <button onClick={() => handleDelete(itemId, price, quantity)}>Delete</button>
    </div>
  )
}

export default CartItem
