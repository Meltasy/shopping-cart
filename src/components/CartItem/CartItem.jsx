const CartItem = ({ itemId, price, quantity, productItems, handleDelete }) => {
  const item = productItems.find(item => item.id === itemId)
  const totalCost = price * quantity
  if (item == null) return null

  return (
    <div key={item.id} className='itemCart'>
      <div className='details'>
        <p>{item.title}</p>
        <h4>€{price ? price.toFixed(2) : '0.00'}</h4>
        <h3>x {quantity}</h3>
        <h4>€{totalCost ? totalCost.toFixed(2): '0.00'}</h4>
      </div>
      <img src={item.image} alt=' ' />
      <button onClick={() => handleDelete(itemId, price, quantity)}>X</button>
    </div>
  )
}

export default CartItem
