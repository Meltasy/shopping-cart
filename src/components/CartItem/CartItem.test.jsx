import { vi, describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react'
import CartItem from './CartItem';
import userEvent from '@testing-library/user-event';

const mockProductItems = [
  { id: '123', title: 'Fake Product A', image: 'fakeImageA.jpg' },
  { id: '321', title: 'Fake Product B', image: 'fakeImageB.jpg' },
]

describe('Cart item component', () => {
  it('renders cart product items', () => {
    const { container } = render(<CartItem itemId='123' price={35.50} quantity={3} productItems={mockProductItems}/>)
    expect(container).toMatchSnapshot()
  })

  it('renders correct cart item details when found', () => {
    render(<CartItem itemId='123' price={35.50} quantity={3} productItems={mockProductItems}/>)
    screen.debug()
    expect(screen.getByText('Fake Product A')).toBeInTheDocument()
    expect(screen.getAllByRole('heading', { hidden: true, level: 4 })[0]).toHaveTextContent('€35.50')
    expect(screen.getByRole('heading', { hidden: true, level: 3 })).toHaveTextContent('3')
    expect(screen.getAllByRole('heading', { hidden: true, level: 4 })[1]).toHaveTextContent('€106.50')
    expect(screen.getByRole('img')).toHaveAttribute('src', 'fakeImageA.jpg')
  })

  it('does not render cart item details when not found', () => {
    render(<CartItem itemId='231' price={35.50} quantity={3} productItems={mockProductItems}/>)
    expect(screen.queryByText('€35.50')).not.toBeInTheDocument()
    expect(screen.queryByText('3')).not.toBeInTheDocument()
  })
})

describe('Cart item component calls handleDelete as expected',() => {
  it('renders delete button with the correct text', () => {
    render(<CartItem itemId='123' price={35.50} quantity={3} productItems={mockProductItems}/>)
    expect(screen.getByRole('button').textContent).toMatch('X')
  })
    
  it('calls handleDelete with correct arguments when button clicked', async () => {
    const mockDelete = vi.fn()
    render(<CartItem itemId='123' price={35.50} quantity={3} productItems={mockProductItems} handleDelete={mockDelete} />)
    const user = userEvent.setup()
    const deleteBtn = screen.getByRole('button', { name: 'X'})
    await user.click(deleteBtn)
    expect(mockDelete).toHaveBeenCalledWith('123', 35.50, 3)
  })
})
