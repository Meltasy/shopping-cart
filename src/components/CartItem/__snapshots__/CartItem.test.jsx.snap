// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart item component > renders cart product items 1`] = `
<div>
  <div
    class="itemCart"
  >
    <div
      class="details"
    >
      <p>
        Fake Product A
      </p>
      <h4>
        €
        35.50
      </h4>
      <h3>
        x 
        3
      </h3>
      <h4>
        €
        106.50
      </h4>
    </div>
    <img
      alt=" "
      src="fakeImageA.jpg"
    />
    <button>
      X
    </button>
  </div>
</div>
`;
